bson/src/main/org/bson/internal/CodecCache.java

```diff
@@ -65,29 +64,18 @@ public String toString() {
         }
     }
 
-    private final ConcurrentMap<CodecCacheKey, Optional<Codec<?>>> codecCache = new ConcurrentHashMap<>();
+    private final ConcurrentMap<CodecCacheKey, Codec<?>> codecCache = new ConcurrentHashMap<>();
 
-    public boolean containsKey(final CodecCacheKey codecCacheKey) {
-        return codecCache.containsKey(codecCacheKey);
+    public <T> Codec<T> putIfAbsent(final CodecCacheKey codecCacheKey, final Codec<T> codec) {
+        assertNotNull(codec);
+        @SuppressWarnings("unchecked")
+        Codec<T> prevCodec = (Codec<T>) codecCache.putIfAbsent(codecCacheKey, codec);
+        return prevCodec == null ? codec : prevCodec;
     }
 
-    public void put(final CodecCacheKey codecCacheKey, final Codec<?> codec){
-        codecCache.put(codecCacheKey, Optional.ofNullable(codec));
-    }
-
-    @SuppressWarnings("unchecked")
-    public synchronized <T> Codec<T> putIfMissing(final CodecCacheKey codecCacheKey, final Codec<T> codec) {
-        Optional<Codec<?>> cachedCodec = codecCache.computeIfAbsent(codecCacheKey, clz -> Optional.of(codec));
-        if (cachedCodec.isPresent()) {
-            return (Codec<T>) cachedCodec.get();
-        }
-        codecCache.put(codecCacheKey, Optional.of(codec));
-        return codec;
-    }
```

Comment:
Instead of improving this method as I originally described [here](https://github.com/mongodb/mongo-java-driver/pull/1007#discussion_r987300328), I did more changes, further simplifying the cache. Before this PR we were caching failures of finding a codec as `Optional.empty()` values just for the sake of throwing an exception faster if asked for the same codec again. I see no point in trying to optimize that path.